NetKet/Machine/pymachine.hpp

```diff
@@ -116,49 +112,45 @@ void AddMachineModule(py::module &m) {
-        subm, "Jastrow")
-        .def(py::init<std::shared_ptr<const AbstractHilbert>>(),
+    py::class_<DerMachine, AbMachineType>(subm, "Jastrow")
+        .def(py::init<const AbstractHilbert &>(), py::keep_alive<1, 2>(),
              py::arg("hilbert")) ADDMACHINEMETHODS(DerMachine);
   }
 
   {
     using DerMachine = JastrowSymm<MachineType>;
-    py::class_<DerMachine, AbMachineType, std::shared_ptr<DerMachine>>(
-        subm, "JastrowSymm")
-        .def(py::init<std::shared_ptr<const AbstractHilbert>>(),
+    py::class_<DerMachine, AbMachineType>(subm, "JastrowSymm")
+        .def(py::init<const AbstractHilbert &>(), py::keep_alive<1, 2>(),
              py::arg("hilbert")) ADDMACHINEMETHODS(DerMachine);
   }
 
 #ifndef COMMA
 #define COMMA ,
 #endif
-  py::class_<MPSPeriodic<MachineType, true>, AbMachineType,
-             std::shared_ptr<MPSPeriodic<MachineType, true>>>(
+  py::class_<MPSPeriodic<MachineType, true>, AbMachineType>(
       subm, "MPSPeriodicDiagonal")
-      .def(py::init<std::shared_ptr<const AbstractHilbert>, double, int>(),
-           py::arg("hilbert"), py::arg("bond_dim"), py::arg("symperiod") = -1)
+      .def(py::init<const AbstractHilbert &, double, int>(),
+           py::keep_alive<1, 2>(), py::arg("hilbert"), py::arg("bond_dim"),
+           py::arg("symperiod") = -1)
           ADDMACHINEMETHODS(MPSPeriodic<MachineType COMMA true>);
 
-  py::class_<MPSPeriodic<MachineType, false>, AbMachineType,
-             std::shared_ptr<MPSPeriodic<MachineType, false>>>(subm,
-                                                               "MPSPeriodic")
-      .def(py::init<std::shared_ptr<const AbstractHilbert>, double, int>(),
-           py::arg("hilbert"), py::arg("bond_dim"), py::arg("symperiod") = -1)
+  py::class_<MPSPeriodic<MachineType, false>, AbMachineType>(subm,
+                                                             "MPSPeriodic")
+      .def(py::init<const AbstractHilbert &, double, int>(),
+           py::keep_alive<1, 2>(), py::arg("hilbert"), py::arg("bond_dim"),
+           py::arg("symperiod") = -1)
           ADDMACHINEMETHODS(MPSPeriodic<MachineType COMMA false>);
 
   AddActivationModule(m);
   AddLayerModule(m);
 
   {
     using DerMachine = FFNN<MachineType>;
-    py::class_<DerMachine, AbMachineType, std::shared_ptr<DerMachine>>(subm,
-                                                                       "FFNN")
-        .def(py::init<
-                 std::shared_ptr<const AbstractHilbert>,
-                 std::vector<std::shared_ptr<AbstractLayer<MachineType>>> &>(),
-             py::arg("hilbert"), py::arg("layers"))
-            ADDMACHINEMETHODS(DerMachine);
+    py::class_<DerMachine, AbMachineType>(subm, "FFNN")
+        .def(py::init<const AbstractHilbert &,
+                      const std::vector<AbstractLayer<MachineType> *> &>(),
+             py::keep_alive<1, 2>(), py::keep_alive<1, 3>(), py::arg("hilbert"),
```

Comment:
> On second thought, it looks like the keep_alive policy refers to python objects, not C++ objects. So this should be okay.

Yes. In my understanding, this keeps the Python `list` object alive which in turn will keep alive the layers it contains. This could still pose a problem if the list is modified, e.g.,  this code causes a segfault:
```python
layers = [...]
m = nk.machine.FFNN(..., layers)
layers.pop(0)
# use the machine...
```
Not sure if there is an easy way to have the machine keep alive all layers in the list directly.
One option would be to go for an `add_layer` method, like `Vmc.add_observable`, instead of passing a list to the constructor.

Reply:
Why do you think it's better by value? move constructor?

Reply:
OK I see the point